### PR TITLE
KOGITO-593: Add pagination controls to datalist, load more button 

### PR DESCRIPTION
**Pull request by nmirasch**

### CODE REVIEW

**Comment by nmirasch** [comment]
@AjayJagan @cristianonicolai @srambach With pending things like remove get all process instances query and provide some design to the load more button, here is the first approach for the load more functionality, just to see if this is in the right way. Any comment, thought, as always, is very welcome, Thanks!

**Comment by AjayJagan** [comment]
> @AjayJagan @cristianonicolai @srambach With pending things like remove get all process instances query and provide some design to the load more button, here is the first approach for the load more functionality, just to see if this is in the right way. Any comment, thought, as always, is very welcome, Thanks!

@nmirasch - looks really cool apart from the pending things that you suggested. :+1:

**Comment by kywalker-rh** [comment]
@nmirasch: My proposal for loading more instances is here: https://marvelapp.com/5jj2d42. Feel free to comment there or here if you have any issues with it. Thanks!

**Comment by nmirasch** [comment]
@kywalker-rh looks good to me, let me give it a try. Thanks!

**Comment by kywalker-rh** [comment]
@nmirasch I added some more detail about my thoughts on how to load the instances into the Jira issue, see: https://issues.redhat.com/browse/KOGITO-1281

**Comment by nmirasch** [comment]
@AjayJagan @Sara4994 @cristianonicolai updated design with @kywalker-rh  suggestions. Could you take a look? Thanks!

File: packages/management-console/src/components/Atoms/LoadMoreComponent/LoadMoreComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+const LoadMoreComponent: React.FC<IOwnProps> = ({
+  offset,
+  setOffset,
+  getProcessInstances,
+  pageSize
+}) => {
+  const loadMore = () => {
+    // setIsLoading(true);
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Atoms/LoadMoreComponent/LoadMoreComponent.tsx]: can we remove :)

File: packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx [hunk 2]
```diff
@@ -21,14 +22,22 @@ interface IOwnProps {
   setIsStatusSelected: any;
   filters: any;
   setFilters: any;
+  setOffset: any;
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx]: number?

File: packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx [hunk 3]
```diff
@@ -21,14 +22,22 @@ interface IOwnProps {
...
+  getProcessInstances: any;
+  setLimit: any;
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx]: number?

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 4]
```diff
@@ -97,6 +104,13 @@ const DataListComponent: React.FC<IOwnProps> = ({
             body="Try using different filters"
           />
         )}
+      {(isLoadingMore) && (
+         <DataListCell className="kogito-management-console-load-more">
+          <Button  variant="secondary">
+              Loading..<Spinner size="md"/>
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]:
```suggestion
              Loading...<Spinner size="md"/>
```

File: packages/management-console/src/graphql/graphql.schema.json [hunk 5]
```diff
@@ -153,6 +153,147 @@
             },
             "isDeprecated": false,
             "deprecationReason": null
+          },
+          {
+            "name": "VisaApplications",
```
Comment (cristianonicolai) [inline packages/management-console/src/graphql/graphql.schema.json]: @nmirasch looks like you generated the schema using data index with custom types deployed. I think we should avoid having those in here.

File: packages/management-console/src/graphql/types.tsx [hunk 6]
```diff
@@ -6,84 +6,182 @@ export type Maybe<T> = T | null;
...
+};
+
+export type AddressArgument = {
+  city?: Maybe<StringArgument>,
+  country?: Maybe<StringArgument>,
+  street?: Maybe<StringArgument>,
+  zipCode?: Maybe<StringArgument>,
+};
+
+export type AddressOrderBy = {
```
Comment (cristianonicolai) [inline packages/management-console/src/graphql/types.tsx]: Same here, custom types were added.

**Review by cristianonicolai — COMMENTED** [review]
@nmirasch overall looks good and works well.
I added a few comments, please also take a look in the code formatting.

Regarding the styling, can we make the row from the load more to have the same height as the other ones? Aligning the button at the middle.
Selection_035

File: packages/management-console/src/components/Atoms/LoadMoreComponent/LoadMoreComponent.tsx (hunk 1 above)
Comment (nmirasch) [reply packages/management-console/src/components/Atoms/LoadMoreComponent/LoadMoreComponent.tsx]: done :)

File: packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx (hunk 2 above)
Comment (nmirasch) [reply packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx]: I'm passing the function setOffset not the offset value, I think this have to be 'any' here

File: packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx (hunk 3 above)
Comment (nmirasch) [reply packages/management-console/src/components/Molecules/DataToolbarComponent/DataToolbarComponent.tsx]: the same here, passing functions not numbers..

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx (hunk 4 above)
Comment (nmirasch) [reply packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: done!

File: packages/management-console/src/graphql/graphql.schema.json (hunk 5 above)
Comment (nmirasch) [reply packages/management-console/src/graphql/graphql.schema.json]: yes, absolute right. Regenerated!

**Comment by nmirasch** [comment]
@cristianonicolai I added some styling changes to load button area to have the same height as the others, and tried to put the button in the middle. Could you give this another try? Thanks!

**Comment by cristianonicolai** [comment]
@nmirasch I'm seeing the load more button when loading spinner is in place. Simply click on Home in the breadcrumb to predroduce.
Screenshot from 2020-03-13 01-35-10
 
@nmirasch @srambach @kywalker-rh  I'm wondering if we should include a "number of items items" to load. A select in the toolbar for user to select how many items to be loaded at time.
Selection_036

Also, should we scroll down automatically after new items are loaded?

**Comment by kywalker-rh** [comment]
@cristianonicolai I like the ability for users to select how many are loaded at a time, if you can implement that in a way that is easy to understand I'm all for it.

And when loaded (especially since the Load more button is at the bottom of the current list) I think the list should stay where it is. The scroll bar size change and the Load more button disappearing will be enough of a cue that the list is now longer.

**Comment by cristianonicolai** [comment]
@kywalker-rh cool, thanks for the feedback. Any suggestions on how it would be placed in that toolbar? The pagination table uses the number of items, but we don't have that in here. The description also needs to be different as we don't have pages.
@nmirasch @AjayJagan @srambach ^

**Comment by nmirasch** [comment]
@cristianonicolai @srambach @AjayJagan  for me just Items to show would work, if you want to see more just click 'load more' :)
The question is where to let to configure that. Let me try to add there in this pr a 'Items to show' selector to the toolbar and see if it fits, something similar to the button to sort here https://www.patternfly.org/v4/documentation/react/demos/toolbardemo

**Comment by kywalker-rh** [comment]
@cristianonicolai and @nmirasch: Given that the Load more doesn't show up except at the bottom of the list maybe there is a way to implement the number of items to load there? I'll work on a quick mock-up to see if I can design something that looks reasonable.

**Comment by kywalker-rh** [comment]
@cristianonicolai @nmirasch: Thanks to @srambach for the idea! Simply having a few buttons to give the user a choice instead of one button looks like the best option to me: 
Screen Shot 2020-03-13 at 10 27 13 AM

Hopefully its fairly easy to implement too.

**Comment by nmirasch** [comment]
@cristianonicolai @AjayJagan @srambach @kywalker-rh  I just added this approach to the pr. Replacing the 'Load all' by 'Load 100 more' :) . I keep the 'Loading...' as it was, one button with the spinner, does it makes sense now, I mean with this new 4 buttons there?

**Comment by cristianonicolai** [comment]
@nmirasch looks good, just need to review a snapshot to get a green build.

**Comment by kywalker-rh** [comment]
Sounds good to me @nmirasch. Thanks!

**Comment by AjayJagan** [comment]
> @cristianonicolai @AjayJagan @srambach @kywalker-rh I just added this approach to the pr. Replacing the 'Load all' by 'Load 100 more' :) . I keep the 'Loading...' as it was, one button with the spinner, does it makes sense now, I mean with this new 4 buttons there?

Great! Looks fine to me.

**Comment by nmirasch** [comment]
@cristianonicolai @AjayJagan @srambach it's green! Could you review? Thanks!

**Review by cristianonicolai — APPROVED** [review]

File: packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx [hunk 7]
```diff
@@ -98,6 +105,13 @@ const DataListComponent: React.FC<IOwnProps> = ({
             body="Try using different filters"
           />
         )}
+      {(isLoadingMore) && (
+         <DataListCell className="kogito-management-console-load-more">
+          <Button  variant="secondary">
+              Loading...<Spinner size="md"/>
```
Comment (srambach) [inline packages/management-console/src/components/Organisms/DataListComponent/DataListComponent.tsx]: Literally if you could put one space between that would be awesome lol. Usually I'd add a spacer utility class but that's not working, so a space will do fine.
`Loading... <Spinner size="md"/>`

**Review by srambach — APPROVED** [review]
Just the one minor thing if you would. 🥇 Nice job!

**Review by AjayJagan — APPROVED** [review]
everything seems to work fine :)

**Comment by nmirasch** [comment]
@cristianonicolai @AjayJagan @Sara4994 I found an issue with this code, the buttons are shown when the filter is being changed..   I'm fixing it right now. Please wait for merging

**Review by Sara4994 — APPROVED** [review]

**Comment by Sara4994** [comment]
> @cristianonicolai @AjayJagan @Sara4994 I found an issue with this code, the buttons are shown when the filter is being changed.. I'm fixing it right now. Please wait for merging

Yep, Just noticed the same

**Review by Sara4994 — APPROVED** [review]

**Pull request merged**